controller_manager/src/controller_manager.cpp

```diff
@@ -53,61 +58,787 @@ ControllerManager::load_controller(
+  while (rclcpp::ok() && switch_params_.do_switch) {
+    if (!rclcpp::ok()) {
+      return controller_interface::return_type::ERROR;
+    }
+    std::this_thread::sleep_for(std::chrono::microseconds(100));
+  }
+  start_request_.clear();
+  stop_request_.clear();
+
+  RCLCPP_DEBUG(get_logger(), "Successfully switched controllers");
+  return controller_interface::return_type::SUCCESS;
+}
+
+controller_interface::ControllerInterfaceSharedPtr
 ControllerManager::add_controller_impl(
-  std::shared_ptr<controller_interface::ControllerInterface> controller,
-  const std::string & controller_name)
+  const ControllerSpec & controller)
+{
+  // lock controllers
+  std::lock_guard<std::recursive_mutex> guard(rt_controllers_wrapper_.controllers_lock_);
+
+  std::vector<ControllerSpec> & to = rt_controllers_wrapper_.get_unused_list(guard);
+  const std::vector<ControllerSpec> & from = rt_controllers_wrapper_.get_updated_list(guard);
+  to.clear();
+
+  // Copy all controllers from the 'from' list to the 'to' list
+  for (const auto & from_controller : from) {
+    to.push_back(from_controller);
+  }
+
+  // Checks that we're not duplicating controllers
+  for (const auto & to_controller : to) {
+    if (controller.info.name == to_controller.info.name) {
+      to.clear();
+      RCLCPP_ERROR(
+        get_logger(),
+        "A controller named '%s' was already loaded inside the controller manager",
+        controller.info.name.c_str());
+      return nullptr;
+    }
+  }
+
+  controller.c->init(hw_, controller.info.name);
+
+  // TODO(v-lopez) this should only be done if controller_manager is configured.
+  // Probably the whole load_controller part should fail if the controller_manager
+  // is not configured, should it implement a LifecycleNodeInterface
+  // https://github.com/ros-controls/ros2_control/issues/152
+  controller.c->get_lifecycle_node()->configure();
+  executor_->add_node(controller.c->get_lifecycle_node()->get_node_base_interface());
+  to.emplace_back(controller);
+
+  // Destroys the old controllers list when the realtime thread is finished with it.
+  RCLCPP_DEBUG(get_logger(), "Realtime switches over to new controller list");
+  rt_controllers_wrapper_.switch_updated_list(guard);
+  RCLCPP_DEBUG(get_logger(), "Destruct controller");
+  std::vector<ControllerSpec> & new_unused_list = rt_controllers_wrapper_.get_unused_list(
+    guard);
+  new_unused_list.clear();
+  RCLCPP_DEBUG(get_logger(), "Destruct controller finished");
+
+  return to.back().c;
+}
+
+controller_interface::ControllerInterfaceSharedPtr ControllerManager::get_controller_by_name(
+  const std::string & name)
+{
+  // lock controllers
+  std::lock_guard<std::recursive_mutex> guard(rt_controllers_wrapper_.controllers_lock_);
+
+  for (const auto & controller : rt_controllers_wrapper_.get_updated_list(guard)) {
+    if (controller.info.name == name) {
+      return controller.c;
+    }
+  }
+  return nullptr;
```

Comment:
`std::find_if`?

Reply:
Done

Reply:
is it?

Reply:
No, I missed this one. Sorry about that.

In the end I removed the function in https://github.com/ros-controls/ros2_control/pull/139/commits/2fba683d2fb8b46fae1a37e66bcdba6acfacd543.